Add view_as_real, view_as_complex for complex tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39860 Setter for real and imag tensor attributes
* **#39099 Add view_as_real, view_as_complex for complex tensors**

Differential Revision: [D22057886](https://our.internmc.facebook.com/intern/diff/D22057886)

This PR adds the following:
1. `view_as_real`, `view_as_complex` for complex tensors
2. Adds the replay logic for these complex views that cause a change in the metadata (in this case- dtype):
`as_strided` doesn't carry the tensor metadata information so this PR adds a replay logic similar to what `XLA` does for these view functions that cause a change in metadata. There are four possible scenarios concerning this PR:
1.  view function that uses as_strided followed by a view function that uses replay (example. `.transpose(...)` followed by `view_as_real`)
2. view function that uses replay followed by a view function that uses as_strided (example. `view_as_real` followed by `.transpose(...)`)
3. view function that uses replay followed by another view function that uses replay (example. `view_as_real` followed by `view_as_complex`)
4.  view function that uses as_strided followed by another view function that uses as_strided (this is what always happens for non-xla tensors rn)

this covers case 1: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152R130

this covers case 2: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152R154

Case 3. is handled by existing XLA logic here: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152L120

anjali411 committed:
Add view_as_real for complex tensors

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add tests
3. autograd code

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add autograd tests

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add autograd tests

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -13,17 +13,50 @@ inline std::vector<int64_t> computeStrideForComplex(IntArrayRef oldstride) {
   // last dimension will always have two elements containing the real and imag vals
   new_sizes.emplace_back(2);
-  auto new_strides = computeStrideForComplex(self.strides());
+  auto strides = self.strides().vec();
```
nit: you don't need the `.vec()` here since you already do it inside the `computeStrideForViewAsReal` function right?

anjali411 replied:
yeah removed the line

albanD commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -13,17 +13,50 @@ inline std::vector<int64_t> computeStrideForComplex(IntArrayRef oldstride) {
+  auto res = oldstride.vec();
+  int dim = res.size();
+
+  TORCH_INTERNAL_ASSERT(res[dim-1] == 1 || res[dim-1] == 0);
```
In the case where `res[dim-1] == 0` you don't need the division below right?

anjali411 replied:
sorry I don't follow. `res[dim -1]` is being removed from the tensor everytime

albanD replied:
But if the stride is 0, your conversion to complex will fail right. Because only a single floating point number is in that dimension, not two as you expect.

anjali411 replied:
I don't think so because you could have tensors of some arbitrary shape just viewing one value and you should still be able to view that tensor as complex or real. for eg. `grad` is initialized in a similar way

albanD commented on test/test_torch.py:
```diff
@@ -17768,6 +17767,45 @@ def is_view_of(self, base, other):
 
+    @onlyOnCPUAndCUDA
+    def test_views_with_dtype_change(self, device):
+        def do_transpose(x, contiguous=False, dim0=0, dim1=1):
```
This function looks a bit weird. In particular because it is called `do_transpose` but does not always do transpose... 
Maybe find a better name?

albanD commented on test/test_torch.py:
```diff
@@ -17786,21 +17824,22 @@ def test_real_imag_view(self, device, dtype):
-            self.assertTrue(t.storage().data_ptr() == re.storage().data_ptr())
+            # for the case of contiguous_input, t=u
+            # for the case of non contiguous_input, the base still remains
+            # u since we are performing a view operation to make the input non-contiguous
```
You mean that the base is t here right?

anjali411 replied:
yup updated

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -57,13 +57,17 @@
     'sparse_coo_tensor_with_dims_and_tensors': 'values',
 }
 
+VIEW_FUNCTIONS_WITH_DTYPE_CHANGE = ['view_as_real', 'view_as_complex']
```
I think this is much more general than dtype change.
It should be all functions for which as_strided is not a good backward.

anjali411 replied:
hmm agreed but what are some other categories of functions for which as_strided would not be a good backward?

albanD replied:
I guess any function that changes metadata. Viewing between sparse and regular Tensors for example.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -111,18 +111,29 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
-        };
+        if (prev_fn == nullptr) {
+          view_func = [=](const at::Tensor& root_base) {
+            auto sizes = root_base.sizes().vec();
```
Do you actually need to make copies with the `.vec()` here?

anjali411 replied:
no you're right updated

albanD reviewed (commented):
I did a first pass.
I'll take another look when all the tests are in.

anjali411 commented on c10/core/ScalarType.h:
```diff
@@ -320,6 +320,17 @@ static inline ScalarType toValueType(ScalarType t) {
+    case ScalarType::Double:
+      return ScalarType::ComplexDouble;
+    default:
+      return t;
```
yeah good catch! updated

anjali411 commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -111,18 +111,29 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
-          auto temp = prev_fn(root_base);
-          return fn(temp);
-        };
+        if (prev_fn == nullptr) {
```
yeah I miswrote it earlier. updated the code

anjali411 commented on tools/autograd/gen_variable_type.py:
```diff
@@ -249,9 +249,9 @@
+SETUP_REPLAY_VIEW_IF_NOT_SUPPORT_AS_STRIDED_OR_VIEW_WITH_DTYPE_CHANGE = CodeTemplate("""\
 c10::optional<std::function<at::Tensor(const at::Tensor&)>> func=c10::nullopt;
-if (!self.unsafeGetTensorImpl()->support_as_strided()) {
+if (!self.unsafeGetTensorImpl()->support_as_strided() || ${is_view_with_dtype_change}) {
```
yeah good point! updated

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add autograd tests

[ghstack-poisoned]

anjali411 commented:
codegen for `view_as_real`:
```
Tensor view_as_real(const Tensor & self) {
  auto& self_ = unpack(self, "self", 0);
  std::shared_ptr<ViewAsRealBackward> grad_fn;
  if (compute_requires_grad( self )) {
    grad_fn = std::shared_ptr<ViewAsRealBackward>(new ViewAsRealBackward(), deleteNode);
    grad_fn->set_next_edges(collect_next_edges( self ));
  }
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto tmp = ([&]() {
    at::AutoNonVariableTypeMode non_var_type_mode(true);
    return at::view_as_real(self_);
  })();
  c10::optional<std::function<at::Tensor(const at::Tensor&)>> func=c10::nullopt;
  if (true || !self.unsafeGetTensorImpl()->support_as_strided()) {
    func = [=](const at::Tensor& input_base) {
      return at::view_as_real(input_base);
    };
  }
  auto result = as_view(/* base */ self, /* output */ tmp, /* is_differentiable */ true, /* view_func */ func, /* creation_meta */ GradMode::is_enabled() ? CreationMeta::DEFAULT: CreationMeta::NO_GRAD_MODE);
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  if (grad_fn) {
      set_history(flatten_tensor_args( result ), grad_fn);
  }
  return result;
}
```

anjali411 commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -111,18 +111,29 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
   auto base_var = Variable(base);
   if (base_var.is_view()) {
```
I updated the code and left comments for three cases (either one or both of current_view and parent has view_func), and the fourth case is the current default behavior for non-xla tensors

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add autograd tests

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


[WIP]
TODO: 
1. add documentation
2. add autograd tests

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


TODO: gradcheck for all complex functions (holomorphic or not)

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


TODO: gradcheck for all complex functions (holomorphic or not)

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


TODO: gradcheck for all complex functions (holomorphic or not)

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -111,33 +111,56 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
+            auto sizes = root_base.sizes();
+            auto strides = root_base.strides();
+            auto storage_offset = root_base.storage_offset();
+            auto temp = root_base.as_strided(sizes, strides, storage_offset);
```
This isn't doing anything. You just view a Tensor with it's own size and stride.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -111,33 +111,56 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
+        auto sizes = temp.sizes();
+        auto strides = temp.strides();
+        auto storage_offset = temp.storage_offset();
+        return temp.as_strided(sizes, strides, storage_offset);
```
Same as above. This is a no-op

albanD commented on test/test_autograd.py:
```diff
@@ -1502,6 +1502,30 @@ def func(x):
         gradcheck(func, [x])
         gradgradcheck(func, [x])
 
+    def test_complex_view_functions(self):
```
This is just testing the backward. not the views.

anjali411 replied:
yeah that's correct. This file just contains autograd specific tests so the name was just to reference which functions' backwards are being tested here.

albanD commented on test/test_torch.py:
```diff
@@ -17923,6 +17922,46 @@ def is_view_of(self, base, other):
         return True
 
+    @onlyOnCPUAndCUDA
+    def test_views_with_dtype_change(self, device):
```
this is just testing view_as_real and view_as_complex right? 
maybe split this in two tests since you don't reuse code?

albanD reviewed (commented):
I think there are a few missing tests:
- tests for all the cases of view combinations for the graph rewrite. keep in mind that you need to do an inplace op on the view to trigger the use of the view_fn, otherwise it does not use it.
- test for gradients: can these just be added to the set of generic tests?

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 commented on torch/_torch_docs.py:
```diff
@@ -1114,22 +1114,90 @@ def merge_dicts(*dicts):
+
+Returns a view of :attr:`input` as a complex tensor. For an input complex tensor of
+:attr:`size` :math:`m1, m2, \dots, mi`, this function returns a new
+real tensor of size :math:`m1, m2, \dots, mi x 2`, where the last dimension of size 2
```
yeah didn't realize

anjali411 commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -13,17 +13,49 @@ inline std::vector<int64_t> computeStrideForComplex(IntArrayRef oldstride) {
+  auto new_strides = computeStrideForViewAsComplex(self.strides());
+
+  const auto complex_type = c10::toComplexType(self.scalar_type());
+  return at::empty({0}, self.options().dtype(complex_type)).set_(self.storage(), self.storage_offset(), new_sizes, new_strides);
```
yeah updated. For all the cases I can think of where the offset is not divisible by 2, I think one of the above torch checks would throw an error. However, to be on the safe side, I still added an assert

albanD replied:
If you do `torch.rand(11).narrow(0, 1, 10).view(5, 2).view_as_complex()` you should get a nice error explaining the issue.

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

albanD commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -53,9 +53,12 @@ Tensor view_as_complex(const Tensor& self) {
   const auto complex_type = c10::toComplexType(self.scalar_type());
-  return at::empty({0}, self.options().dtype(complex_type)).set_(self.storage(), self.storage_offset(), new_sizes, new_strides);
+
+  TORCH_INTERNAL_ASSERT(self.storage_offset() % 2 == 0);
```
This is not an internal assert, a user can trigger it.

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 commented:
> I think there are a few missing tests:
> 
> * tests for all the cases of view combinations for the graph rewrite. keep in mind that you need to do an inplace op on the view to trigger the use of the view_fn, otherwise it does not use it.

added the tests

> * test for gradients: can these just be added to the set of generic tests?

no I just manual tests because `view_as_complex`, `view_as_real`, `real`, `imag` are non-holomorphic functions

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -1502,6 +1502,66 @@ def func(x):
         gradcheck(func, [x])
         gradgradcheck(func, [x])
 
+    def test_complex_views(self):
```
Can you add a couple comments in this function to explain what you're testing and why these are special cased here.

albanD commented on test/test_autograd.py:
```diff
@@ -1502,6 +1502,66 @@ def func(x):
+
+    def test_view_func_for_complex_views(self):
+        x = torch.randn(2, 2, 2, dtype=torch.double, requires_grad=True)
+        y = x.clone().detach().requires_grad_(True)
```
nit: the clone is not really necessary here because you won't be able to modify x inplace anyways.
same for b below

albanD commented on test/test_autograd.py:
```diff
@@ -1502,6 +1502,66 @@ def func(x):
+        func(z).backward()
+        self.assertEqual(z.grad, z1.grad)
+
+    def test_view_func_for_complex_views(self):
```
This might be clearer and more reliable if you separate out the test to test each case one after the other.
Also I think you want to test the following to actually catch errors:

Something like the following (that uses chunk that is currently broken and does not mark the result as views):
```python
import torch

a = torch.rand(10, dtype=torch.double, requires_grad=True)

def fn(a):
    b = a.clone()
    c1 = b.chunk(4)[0]
    c2 = b.chunk(4)[0]

    c1.mul_(2)
    c2.mul_(3)

    return b

torch.autograd.gradcheck(fn, a)
```

anjali411 replied:
so you mean replace `chunk` with `view_as_complex` and `view_as_real` and add a similar test?

albanD replied:
Yes change the way c1 and c2 are created to use the view function you want to test.
This should fail before this PR and work with it.

albanD commented on tools/autograd/gen_autograd.py:
```diff
@@ -64,11 +64,16 @@
+VIEW_FUNCTIONS_WITH_METADATA_CHANGE = ['view_as_real', 'view_as_complex']
+
+for key in VIEW_FUNCTIONS_WITH_METADATA_CHANGE:
+    VIEW_FUNCTIONS[key] = 'self'
```
nit for consistency, I would have done the same as what is done below for `RETURNS_VIEWS_OF_INPUT` and just add the two.

anjali411 replied:
but `real` and `imag` are composite view functions and they are not different from chunk or split in any way so I don't see why we need to do that

albanD replied:
I didn't mean that you should change these lists at all.
It is just that we have 3 lists that contain the previous one. So it would be simpler to just create them in that order.
But we can re-order that later if you prefer.

albanD commented on torch/csrc/autograd/VariableTypeUtils.h:
```diff
@@ -111,33 +111,56 @@ inline Tensor as_view(const Tensor & base, Tensor tensor, bool is_differentiable
-        };
+        // current_view has a view_func and but it's parent doesn't have one
+        if(base_var.unsafeGetTensorImpl()->support_as_strided()) {
+          auto size = tensor.sizes().vec();
```
Don't you want `base` here?
Make sure to add some test that stresses this out.

anjali411 replied:
updated!

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"



[ghstack-poisoned]

albanD commented on c10/core/ScalarType.h:
```diff
@@ -320,6 +320,17 @@ static inline ScalarType toValueType(ScalarType t) {
+    case ScalarType::Double:
+      return ScalarType::ComplexDouble;
+    default:
+      AT_ERROR("Unknown Complex ScalarType");
```
Isn't AT_ERROR deprecated? We should only use `TORCH_CHECK` now right?

anjali411 replied:
oh okay updated

albanD reviewed (commented):
Looks mostly ok.
Missing the definition for the vjp of complex in the autograd notes

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


Differential Revision: [D22057886](https://our.internmc.facebook.com/intern/diff/D22057886)

[ghstack-poisoned]

anjali411 commented:
> Looks mostly ok.
> Missing the definition for the vjp of complex in the autograd notes

@albanD added note for how autograd works for complex

mruberry commented on test/test_autograd.py:
```diff
@@ -4499,6 +4502,105 @@ def fn(*inputs):
             setattr(TestAutogradDeviceType, test_name, do_test)
 
+class TestAutogradComplex(TestCase):
+    # remove this test after gradcheck support is added for non-holomoprphic functions
```
sp: "non-holomoprphic"

mruberry commented on test/test_autograd.py:
```diff
@@ -4321,6 +4321,9 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
 
 separate_complex_tests = ['log', 'log10', 'log1p', 'log2', 'reciprocal', 'tan']
 
+# NOTE: Some non-holomorphic are separately tested in TestAutogradComplex until gradcheck works properly
```
Define "(non-)holomorphic"

anjali411 replied:
https://en.wikipedia.org/wiki/Holomorphic_function

mruberry commented on test/test_autograd.py:
```diff
@@ -4499,6 +4502,105 @@ def fn(*inputs):
 
+class TestAutogradComplex(TestCase):
+    # remove this test after gradcheck support is added for non-holomoprphic functions
+    def test_real(self):
```
Can these tests also check a sum over a couple dimensions to ensure the gradient is not always uniform?

mruberry commented on test/test_autograd.py:
```diff
@@ -4499,6 +4502,105 @@ def fn(*inputs):
+    # remove this test after gradcheck support is added for non-holomoprphic functions
+    def test_view_as_real(self):
+        x = torch.randn(10, dtype=torch.cdouble, requires_grad=True)
+        torch.view_as_real(x).sum().backward()
```
What about a test like this one EXCEPT:

- you backward through torch.view_as_real AND the complex tensor itself, then validate the gradient on the original complex tensor is correct
- you view_as_real, then view_as_complex from the tensor returned from view_as_real, then view_as_real on that then backward and validate
-

mruberry replied:
Maybe the latter case is sufficiently covered by "as_identity" below?

anjali411 replied:
yeah that case is covered by as_identity

mruberry commented on test/test_autograd.py:
```diff
@@ -4499,6 +4502,105 @@ def fn(*inputs):
+        x.real.sum().backward()
+        self.assertEqual(x.grad, torch.ones_like(x))
+
+    # remove this test after gradcheck support is added for non-holomoprphic functions
```
what about a test where you manipulate both the real and imag parts and then backward through both those results?

anjali411 replied:
I don't see a case that this test covers which hasn't already been covered

mruberry commented on test/test_autograd.py:
```diff
@@ -4499,6 +4502,105 @@ def fn(*inputs):
+        torch.view_as_complex(x).sum().backward()
+        self.assertEqual(x.grad, torch.tensor([1, 0], dtype=torch.double).expand_as(x))
+
+    def test_view_func_for_complex_views(self):
```
Is there a test for in-place operations on .real and .imag? No, right, because that will be part of the setter PR?

anjali411 replied:
yeah that's correct

mruberry commented on test/test_torch.py:
```diff
@@ -18021,6 +18020,48 @@ def is_view_of(self, base, other):
 
         return True
 
+    def transpose_if_not_contig(self, x, contiguous=False, dim0=0, dim1=1):
```
_transpose_if_not_contig (helper functions typically start their names with an underscore)

mruberry replied:
Also don't take defaults. If you want to require contiguous, dim, etc. be specified with keywords just put them after a *. Defaults are appropriate for user-facing functions, but I think this is developer facing?

mruberry replied:
This function has a misleading name that suggests it's checking the contiguity of x. Maybe get rid of it and just enumerate `do_transpose in (True, False):` in the tests below?

anjali411 replied:
updated

mruberry commented on test/test_autograd.py:
```diff
@@ -4499,6 +4502,105 @@ def fn(*inputs):
 
             setattr(TestAutogradDeviceType, test_name, do_test)
 
+class TestAutogradComplex(TestCase):
```
Should this class be made device generic so it runs on CPU and CUDA?

anjali411 replied:
autograd tests run both cpu and cuda already

mruberry replied:
These tests won't run on both but that's OK for now.

mruberry commented on test/test_torch.py:
```diff
@@ -18021,6 +18020,48 @@ def is_view_of(self, base, other):
+
+    @onlyOnCPUAndCUDA
+    def test_view_as_complex(self, device):
+        def fn(contiguous_input=True, dim0=0, dim1=1):
```
This test seems a little confused to me. Would it be simpler if it was separated into two tests, one that catches the runtime error and another that verifies the constructed complex tensor and the tensor return from view_as_complex are equivalent?

anjali411 replied:
No I think it's better to have one test to check for correct output and the case where a RuntimeError is thrown, I added a comment to clarify which function call catches the RuntimeError

mruberry commented on test/test_torch.py:
```diff
@@ -18021,6 +18020,48 @@ def is_view_of(self, base, other):
+    @onlyOnCPUAndCUDA
+    def test_view_as_real(self, device):
+        def fn(contiguous_input=True):
+            t = torch.randn(3, 4, 2, device=device)
```
Conceptually it's a little weird to me that this function doesn't start by creating a complex tensor. It could check complex tensors of 0, 1, and 3 dims, for example.

mruberry replied:
Sorry: another special-case would be a complex tensor with zero elements.

anjali411 replied:
updated (with comments)!

mruberry commented on torch/_torch_docs.py:
```diff
@@ -1114,22 +1114,89 @@ def merge_dicts(*dicts):
+.. warning::
+    :func:`view_as_complex` is only supported for tensors with :class:`torch.dtype` ``torch.float64`` and ``torch.float32`.
+    The input is expected to have the last dimension of :attr:`size` 2. In addition, the tensor must have a `stride` of 1
+    for it's last dimension. The strides of all other dimensions must be even numbers.
```
sp: it's -> its

mruberry commented on test/test_torch.py:
```diff
@@ -18021,6 +18020,48 @@ def is_view_of(self, base, other):
+
+            input = self.transpose_if_not_contig(t, contiguous_input, dim0, dim1)
+
+            if (input.size()[-1] != 2 or
```
Should also check for the even stride requirement when not the innermost. Also, is that innermost check correct? What if the innermost dim has stride 0?

anjali411 replied:
yeah good point! added a case which checks that

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


Differential Revision: [D22057886](https://our.internmc.facebook.com/intern/diff/D22057886)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -13,17 +13,52 @@ inline std::vector<int64_t> computeStrideForComplex(IntArrayRef oldstride) {
+// with corresponding real dtype containing the complex values
+// in the last two dimensions
+Tensor view_as_real(const Tensor& self) {
+  TORCH_CHECK(self.is_complex(), "view_as_real is only supported for complex tensors");
```
I guess we could always no op in this case... nahhhh

anjali411 replied:
like just return self?

anjali411 replied:
umm I think we should enable this just for complex at least for now

ezyang commented on aten/src/ATen/native/ComplexHelper.h:
```diff
@@ -13,17 +13,52 @@ inline std::vector<int64_t> computeStrideForComplex(IntArrayRef oldstride) {
   return at::empty({0}, self.options().dtype(float_type)).set_(self.storage(), new_storage_offset, new_sizes, new_strides);
 }
 
+inline std::vector<int64_t> computeStrideForViewAsComplex(IntArrayRef oldstride) {
```
consider using a DimVector here

anjali411 replied:
what's the advantage of using `DimVector`?

ezyang commented on tools/autograd/derivatives.yaml:
```diff
@@ -1055,6 +1049,12 @@
   self: grad.reshape(self.sizes())
 
+- name: view_as_real(Tensor(a) self) -> Tensor(a)
+  self: at::view_as_complex(grad.contiguous()).conj() # gx0 - i gx1
```
BTW, does this imply our imag gradient from before was wrong? Because it seems like going through `view_as_real` to get the imag part would have skipped the negation

anjali411 replied:
I am not sure what you mean by if our imag gradient was wrong. But what's happening here is:

```
f: R^2 -> C
vjp = [gx, -gy] * [[1, 0], [0, 1]]
```

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


Differential Revision: [D22057886](https://our.internmc.facebook.com/intern/diff/D22057886)

[ghstack-poisoned]

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


Differential Revision: [D22057886](https://our.internmc.facebook.com/intern/diff/D22057886)

This PR adds the following:
1. `view_as_real`, `view_as_complex` for complex tensors
2. Adds the replay logic for these complex views that cause a change in the metadata (in this case- dtype):
`as_strided` doesn't carry the tensor metadata information so this PR adds a replay logic similar to what `XLA` does for these view functions that cause a change in metadata. There are four possible scenarios concerning this PR:
1.  view function that uses as_strided followed by a view function that uses replay (example. `.transpose(...)` followed by `view_as_real`)
2. view function that uses replay followed by a view function that uses as_strided (example. `view_as_real` followed by `.transpose(...)`)
3. view function that uses replay followed by another view function that uses replay (example. `view_as_real` followed by `view_as_complex`)
4.  view function that uses as_strided followed by another view function that uses as_strided (this is what always happens for non-xla tensors rn)

this covers case 1: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152R130

this covers case 2: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152R154

Case 3. is handled by existing XLA logic here: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152L120



[ghstack-poisoned]

mruberry reviewed (approved):
Cool! 

Just remember to update docs/source/torch.rst.

anjali411 committed:
Update on "Add view_as_real, view_as_complex for complex tensors"


Differential Revision: [D22057886](https://our.internmc.facebook.com/intern/diff/D22057886)

This PR adds the following:
1. `view_as_real`, `view_as_complex` for complex tensors
2. Adds the replay logic for these complex views that cause a change in the metadata (in this case- dtype):
`as_strided` doesn't carry the tensor metadata information so this PR adds a replay logic similar to what `XLA` does for these view functions that cause a change in metadata. There are four possible scenarios concerning this PR:
1.  view function that uses as_strided followed by a view function that uses replay (example. `.transpose(...)` followed by `view_as_real`)
2. view function that uses replay followed by a view function that uses as_strided (example. `view_as_real` followed by `.transpose(...)`)
3. view function that uses replay followed by another view function that uses replay (example. `view_as_real` followed by `view_as_complex`)
4.  view function that uses as_strided followed by another view function that uses as_strided (this is what always happens for non-xla tensors rn)

this covers case 1: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152R130

this covers case 2: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152R154

Case 3. is handled by existing XLA logic here: https://github.com/pytorch/pytorch/pull/39099/files#diff-88d45e2b30fab214a6aa074fd30eb152L120



[ghstack-poisoned]

albanD reviewed (approved):
Thanks Mike for the update.
Good to go from my side !
